### PR TITLE
Allocate fewer objects

### PR DESCRIPTION
**Pull request by tamird**
cc @hundredwatt

### CODE REVIEW

File: activerecord/lib/active_record/relation/record_fetch_warning.rb [hunk 1]
```diff
@@ -23,23 +23,21 @@ def exec_queries
         end
       end
 
-      ActiveSupport::Notifications.subscribe("sql.active_record") do |*args|
-        payload = args.last
-
-        QueryRegistry.queries << payload[:sql]
+      ActiveSupport::Notifications.subscribe("sql.active_record") do |*, payload|
+        QueryRegistry.queries << payload.fetch(:sql)
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/relation/record_fetch_warning.rb]: Why `fetch`?

File: activerecord/test/cases/finder_test.rb [hunk 2]
```diff
@@ -59,7 +59,7 @@ def test_symbols_table_ref
     Post.where("author_id" => nil)  # warm up
     x = Symbol.all_symbols.count
     Post.where("title" => {"xxxqqqq" => "bar"})
-    assert_equal x, Symbol.all_symbols.count
+    assert_operator(Symbol.all_symbols.count, :<=, x)
```
Comment (rafaelfranca) [inline activerecord/test/cases/finder_test.rb]: Why this change?

File: activerecord/lib/active_record/relation/record_fetch_warning.rb (hunk 1 above)
Comment (tamird) [inline activerecord/lib/active_record/relation/record_fetch_warning.rb]: slightly tighter, since we always expect to have this key

File: activerecord/test/cases/finder_test.rb (hunk 2 above)
Comment (tamird) [inline activerecord/test/cases/finder_test.rb]: ah, it's related to another change i had sitting around - let me back it out

**Comment by rafaelfranca** [comment]
I can see that both commits allocates fewer arrays but it would be really great if we know how many. Could you try to benchmark it?

**Comment by tamird** [comment]
I'd love to - is there any prior art you can point me to?

**Comment by rafaelfranca** [comment]
Maybe @sgrif can help you

**Comment by tgxworld** [comment]
@tamird  This might help. https://github.com/eileencodes/integration_performance_test/blob/master/test/integration/documents_allocations_create_test.rb.

**Comment by tamird** [comment]
@rafaelfranca:

allocation_test.rb:
```ruby
require 'cases/helper'
require 'models/post'

module ActiveRecord
  class AllocationBenchmark < ActiveRecord::TestCase
    fixtures :posts

    def test_allocation
      arrays = []
      where_clauses = []
      from_clauses = []
      hashes = []

      10.times do
        Post.all.to_a

        arrays << ObjectSpace.each_object(Array).count
        where_clauses << ObjectSpace.each_object(Relation::WhereClause).count
        from_clauses << ObjectSpace.each_object(Relation::FromClause).count
        hashes << ObjectSpace.each_object(Hash).count
      end

      puts 'arrays: ' << arrays.each_cons(2).map { |left, right| right - left }.inspect
      puts 'where_clauses: ' << where_clauses.each_cons(2).map { |left, right| right - left }.inspect
      puts 'from_clauses: ' << from_clauses.each_cons(2).map { |left, right| right - left }.inspect
      puts 'hashes: ' << hashes.each_cons(2).map { |left, right| right - left }.inspect
    end
  end
end
```

master (fd94aa4d57a7552ad40c13226caba5b022800326):
```
arrays: [73, 73, 73, 73, 73, 73, 73, 73, 73]
where_clauses: [4, 4, 4, 4, 4, 4, 4, 4, 4]
from_clauses: [2, 2, 2, 2, 2, 2, 2, 2, 2]
hashes: [75, 75, 75, 75, 75, 75, 75, 75, 75]
```
d427e9e7e527aa609c96a5c1450983fb67a29f43:
```
arrays: [73, 73, 73, 73, 73, 73, 73, 73, 73]
where_clauses: [4, 4, 4, 4, 4, 4, 4, 4, 4]
from_clauses: [2, 2, 2, 2, 2, 2, 2, 2, 2]
hashes: [75, 75, 75, 75, 75, 75, 75, 75, 75]
```
c150db9019179b00264c3c6d12320187a90327e6:
```
arrays: [70, 70, 70, 70, 70, 70, 70, 70, 70]
where_clauses: [4, 4, 4, 4, 4, 4, 4, 4, 4]
from_clauses: [2, 2, 2, 2, 2, 2, 2, 2, 2]
hashes: [75, 75, 75, 75, 75, 75, 75, 75, 75]
```
267d7df3e0236b481b33689f5f1d3579367c7115:
```
arrays: [66, 66, 66, 66, 66, 66, 66, 66, 66]
where_clauses: [2, 2, 2, 2, 2, 2, 2, 2, 2]
from_clauses: [1, 1, 1, 1, 1, 1, 1, 1, 1]
hashes: [75, 75, 75, 75, 75, 75, 75, 75, 75]
```
67fd2ec02bcd52cfd8064eba27aeacf7d96d2f8f:
```
arrays: [66, 66, 66, 66, 66, 66, 66, 66, 66]
where_clauses: [2, 2, 2, 2, 2, 2, 2, 2, 2]
from_clauses: [1, 1, 1, 1, 1, 1, 1, 1, 1]
hashes: [75, 75, 75, 75, 75, 75, 75, 75, 75]
```

File: activerecord/lib/active_record/relation/query_methods.rb [hunk 3]
```diff
@@ -53,7 +53,7 @@ def not(opts, *rest)
     Relation::MULTI_VALUE_METHODS.each do |name|
       class_eval <<-CODE, __FILE__, __LINE__ + 1
         def #{name}_values                    # def select_values
-          @values[:#{name}] || []             #   @values[:select] || []
+          @values[:#{name}] ||= []            #   @values[:select] ||= []
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/relation/query_methods.rb]: @sgrif this does make sense to me, but WDYT?

Comment (sgrif) [inline activerecord/lib/active_record/relation/query_methods.rb]: I'd like to see since speed benchmarks, since this will show down creation of new relations with the additional values to copy

Comment (tamird) [inline activerecord/lib/active_record/relation/query_methods.rb]: can you help illustrate the kind of benchmark you'd like to see? just relation chaining?

Comment (sgrif) [inline activerecord/lib/active_record/relation/query_methods.rb]:
```ruby
relation = User.all
relation.load
second_relation = User.all
second_relation.load

Benchmark.ips do |bm|
  bm.report("spawn") { relation.spawn }
  bm.report("merge") { relation.merge(second_relation) }
end
```

File: activerecord/lib/active_record/relation/record_fetch_warning.rb (hunk 1 above)
Comment (sgrif) [inline activerecord/lib/active_record/relation/record_fetch_warning.rb]: This is slower than `[]` https://gist.github.com/jonleighton/3552829

Comment (tamird) [inline activerecord/lib/active_record/relation/record_fetch_warning.rb]: :( will change back

Comment (kaspth) [inline activerecord/lib/active_record/relation/record_fetch_warning.rb]: Just tried that benchmark on 2.2.1 and I get:

```
Calculating -------------------------------------
                 #[]   139.259k i/100ms
              #fetch   132.383k i/100ms
-------------------------------------------------
                 #[]      7.907M (± 8.3%) i/s -     39.271M
              #fetch      6.944M (± 7.9%) i/s -     34.552M
```

It seems the discrepancy has decreased to only around 10-15% and `fetch` has become 2.25 times faster.

Comment (sgrif) [inline activerecord/lib/active_record/relation/record_fetch_warning.rb]: Yes, but it's still slower than `[]` which was my point.

File: activerecord/lib/active_record/relation/query_methods.rb (hunk 3 above)
Comment (tamird) [inline activerecord/lib/active_record/relation/query_methods.rb]: Here's what I went with:
```ruby
require 'cases/helper'
require 'models/post'

require 'benchmark/ips'

module ActiveRecord
  class RecordFetchWarningTest < ActiveRecord::TestCase
    fixtures :posts

    def test_bench
      permutations = (0..3).flat_map { |i| ([true] * i + [false] * (3 - i)).permutation.to_a.uniq }

      Benchmark.ips do |bm|
        permutations.each do |permutation|
          $set_array, $set_clause, $set_hash = *permutation

          relation = Post.all
          relation.load
          second_relation = Post.all
          second_relation.load

          bm.report("spawn: set_array: #{$set_array}, set_clause: #{$set_clause}, set_hash: #{$set_hash}") { relation.spawn }
        end

        bm.compare!
      end

      Benchmark.ips do |bm|
        permutations.each do |permutation|
          $set_array, $set_clause, $set_hash = *permutation

          relation = Post.all
          relation.load
          second_relation = Post.all
          second_relation.load

          bm.report("merge: set_array: #{$set_array}, set_clause: #{$set_clause}, set_hash: #{$set_hash}") { relation.merge(second_relation) }
        end

        bm.compare!
      end
    end
  end
end
```
And I applied this diff:
```diff
diff --git a/activerecord/lib/active_record/relation/query_methods.rb b/activerecord/lib/active_record/relation/query_methods.rb
index 867bc16..0b830a5 100644
--- a/activerecord/lib/active_record/relation/query_methods.rb
+++ b/activerecord/lib/active_record/relation/query_methods.rb
@@ -53,7 +53,12 @@ module ActiveRecord
     Relation::MULTI_VALUE_METHODS.each do |name|
       class_eval <<-CODE, __FILE__, __LINE__ + 1
         def #{name}_values                    # def select_values
-          @values[:#{name}] ||= []            #   @values[:select] ||= []
+          # @values[:#{name}] ||= []            #   @values[:select] ||= []
+          if $set_array
+            @values[:#{name}] ||= []
+          else
+            @values[:#{name}] || []
+          end
         end                                   # end
                                               #
         def #{name}_values=(values)           # def select_values=(values)
@@ -83,7 +88,12 @@ module ActiveRecord
     Relation::CLAUSE_METHODS.each do |name|
       class_eval <<-CODE, __FILE__, __LINE__ + 1
         def #{name}_clause                           # def where_clause
-          @values[:#{name}] ||= new_#{name}_clause   #   @values[:where] ||= new_where_clause
+          # @values[:#{name}] ||= new_#{name}_clause   #   @values[:where] ||= new_where_clause
+          if $set_clause
+            @values[:#{name}] ||= new_#{name}_clause
+          else
+            @values[:#{name}] || new_#{name}_clause
+          end
         end                                          # end
                                                      #
         def #{name}_clause=(value)                   # def where_clause=(value)
@@ -98,7 +108,12 @@ module ActiveRecord
     end
 
     def create_with_value # :nodoc:
-      @values[:create_with] ||= {}
+      # @values[:create_with] ||= {}
+      if $set_hash
+        @values[:create_with] ||= {}
+      else
+        @values[:create_with] || {}
+      end
     end
 
     alias extensions extending_values

```
As you say, there is a slowdown when copying:
```
Calculating -------------------------------------
spawn: set_array: false, set_clause: false, set_hash: false
                        28.367k i/100ms
spawn: set_array: true, set_clause: false, set_hash: false
                        16.709k i/100ms
spawn: set_array: false, set_clause: true, set_hash: false
                        23.045k i/100ms
spawn: set_array: false, set_clause: false, set_hash: true
                        27.183k i/100ms
spawn: set_array: true, set_clause: true, set_hash: false
                        15.163k i/100ms
spawn: set_array: true, set_clause: false, set_hash: true
                        16.910k i/100ms
spawn: set_array: false, set_clause: true, set_hash: true
                        20.742k i/100ms
spawn: set_array: true, set_clause: true, set_hash: true
                        14.802k i/100ms
-------------------------------------------------
spawn: set_array: false, set_clause: false, set_hash: false
                        534.965k (± 7.8%) i/s -      2.666M
spawn: set_array: true, set_clause: false, set_hash: false
                        221.189k (±13.4%) i/s -      1.103M
spawn: set_array: false, set_clause: true, set_hash: false
                        364.490k (± 7.8%) i/s -      1.821M
spawn: set_array: false, set_clause: false, set_hash: true
                        452.526k (± 9.0%) i/s -      2.256M
spawn: set_array: true, set_clause: true, set_hash: false
                        189.651k (±12.2%) i/s -    940.106k
spawn: set_array: true, set_clause: false, set_hash: true
                        212.384k (±14.8%) i/s -      1.032M
spawn: set_array: false, set_clause: true, set_hash: true
                        356.481k (±12.0%) i/s -      1.763M
spawn: set_array: true, set_clause: true, set_hash: true
                        195.456k (±11.6%) i/s -    962.130k

Comparison:
spawn: set_array: false, set_clause: false, set_hash: false:   534964.8 i/s
spawn: set_array: false, set_clause: false, set_hash: true:   452526.5 i/s - 1.18x slower
spawn: set_array: false, set_clause: true, set_hash: false:   364489.5 i/s - 1.47x slower
spawn: set_array: false, set_clause: true, set_hash: true:   356481.4 i/s - 1.50x slower
spawn: set_array: true, set_clause: false, set_hash: false:   221188.5 i/s - 2.42x slower
spawn: set_array: true, set_clause: false, set_hash: true:   212383.9 i/s - 2.52x slower
spawn: set_array: true, set_clause: true, set_hash: true:   195456.0 i/s - 2.74x slower
spawn: set_array: true, set_clause: true, set_hash: false:   189650.9 i/s - 2.82x slower

Calculating -------------------------------------
merge: set_array: false, set_clause: false, set_hash: false
                         1.306k i/100ms
merge: set_array: true, set_clause: false, set_hash: false
                         1.277k i/100ms
merge: set_array: false, set_clause: true, set_hash: false
                         1.571k i/100ms
merge: set_array: false, set_clause: false, set_hash: true
                         1.061k i/100ms
merge: set_array: true, set_clause: true, set_hash: false
                         1.512k i/100ms
merge: set_array: true, set_clause: false, set_hash: true
                         1.354k i/100ms
merge: set_array: false, set_clause: true, set_hash: true
                         1.740k i/100ms
merge: set_array: true, set_clause: true, set_hash: true
                         1.512k i/100ms
-------------------------------------------------
merge: set_array: false, set_clause: false, set_hash: false
                         14.672k (± 7.1%) i/s -     73.136k
merge: set_array: true, set_clause: false, set_hash: false
                         13.485k (±10.2%) i/s -     67.681k
merge: set_array: false, set_clause: true, set_hash: false
                         18.798k (± 6.8%) i/s -     94.260k
merge: set_array: false, set_clause: false, set_hash: true
                         15.079k (± 7.2%) i/s -     75.331k
merge: set_array: true, set_clause: true, set_hash: false
                         16.701k (±13.8%) i/s -     81.648k
merge: set_array: true, set_clause: false, set_hash: true
                         12.918k (±14.9%) i/s -     63.638k
merge: set_array: false, set_clause: true, set_hash: true
                         17.527k (± 9.2%) i/s -     87.000k
merge: set_array: true, set_clause: true, set_hash: true
                         15.934k (± 9.6%) i/s -     80.136k

Comparison:
merge: set_array: false, set_clause: true, set_hash: false:    18797.8 i/s
merge: set_array: false, set_clause: true, set_hash: true:    17527.2 i/s - 1.07x slower
merge: set_array: true, set_clause: true, set_hash: false:    16701.1 i/s - 1.13x slower
merge: set_array: true, set_clause: true, set_hash: true:    15934.1 i/s - 1.18x slower
merge: set_array: false, set_clause: false, set_hash: true:    15079.0 i/s - 1.25x slower
merge: set_array: false, set_clause: false, set_hash: false:    14672.3 i/s - 1.28x slower
merge: set_array: true, set_clause: false, set_hash: false:    13485.2 i/s - 1.39x slower
merge: set_array: true, set_clause: false, set_hash: true:    12917.5 i/s - 1.46x slower
```

How shall we proceed?

Comment (matthewd) [inline activerecord/lib/active_record/relation/query_methods.rb]: Aren't those methods' results read-only? Instead of `.. ||= []`, can we use `.. || SHARED_FROZEN_EMPTY_ARRAY`?

Comment (sgrif) [inline activerecord/lib/active_record/relation/query_methods.rb]: :+1: for @matthewd's suggestion.

Comment (sgrif) [inline activerecord/lib/active_record/relation/query_methods.rb]: We can also change `WhereClause.empty` to memoize the result, since it's immutable.

Comment (tamird) [inline activerecord/lib/active_record/relation/query_methods.rb]: `WhereClause.empty` appears to not be immutable - it's modified when `merge` is called

Comment (sgrif) [inline activerecord/lib/active_record/relation/query_methods.rb]: Merge returns a new where clause.

Comment (tamird) [inline activerecord/lib/active_record/relation/query_methods.rb]: I believe you, but things blow up if I change `new_where_clause` to return a frozen object

Comment (tamird) [inline activerecord/lib/active_record/relation/query_methods.rb]: Ha, it's the memoization.
```
      def referenced_columns
        @referenced_columns ||= begin
          equality_nodes = predicates.select { |n| equality_node?(n) }
          Set.new(equality_nodes, &:left)
        end
      end      def referenced_columns
        @referenced_columns ||= begin
          equality_nodes = predicates.select { |n| equality_node?(n) }
          Set.new(equality_nodes, &:left)
        end
      end
```

Comment (sgrif) [inline activerecord/lib/active_record/relation/query_methods.rb]: Do this:

```
result = WhereClause.empty
result.send(:referenced_columns) # This method is memoizes, which will blow up if you freeze it
result.freeze
```

**Comment by tamird** [comment]
Well, I pushed a commit to implement @matthewd's suggestion, and then introduced some instrumentation to benchmark...the results are puzzling.

bench.rb:
```ruby
require 'cases/helper'
require 'models/post'

require 'benchmark/ips'

module ActiveRecord
  class RecordFetchWarningTest < ActiveRecord::TestCase
    fixtures :posts

    def test_bench
      num = 3
      permutations = (0..num).flat_map { |i| ([true] * i + [false] * (num - i)).permutation.to_a.uniq }

      Benchmark.ips do |bm|
        permutations.each do |permutation|
          $set_array, $set_clause, $set_hash = *permutation

          relation = Post.all
          relation.load
          second_relation = Post.all
          second_relation.load

          bm.report("spawn: set_array: #{$set_array}, set_clause: #{$set_clause}, set_hash: #{$set_hash}") { relation.spawn }
        end

        bm.compare!
      end

      Benchmark.ips do |bm|
        permutations.each do |permutation|
          $set_array, $set_clause, $set_hash = *permutation

          relation = Post.all
          relation.load
          second_relation = Post.all
          second_relation.load

          bm.report("merge: set_array: #{$set_array}, set_clause: #{$set_clause}, set_hash: #{$set_hash}") { relation.merge(second_relation) }
        end

        bm.compare!
      end
    end
  end
end
```
Minor diff:
```diff
diff --git a/activerecord/lib/active_record/relation/from_clause.rb b/activerecord/lib/active_record/relation/from_clause.rb
index ee2fc4e..597307f 100644
--- a/activerecord/lib/active_record/relation/from_clause.rb
+++ b/activerecord/lib/active_record/relation/from_clause.rb
@@ -25,7 +25,11 @@ module ActiveRecord
       end
 
       def self.empty
+        if $set_clause
           @empty ||= new(nil, nil).freeze
+        else
+          new(nil, nil)
+        end
       end
     end
   end
diff --git a/activerecord/lib/active_record/relation/query_methods.rb b/activerecord/lib/active_record/relation/query_methods.rb
index 0416cb8..b349bb4 100644
--- a/activerecord/lib/active_record/relation/query_methods.rb
+++ b/activerecord/lib/active_record/relation/query_methods.rb
@@ -54,7 +54,11 @@ module ActiveRecord
     Relation::MULTI_VALUE_METHODS.each do |name|
       class_eval <<-CODE, __FILE__, __LINE__ + 1
         def #{name}_values                        # def select_values
+          if $set_array
             @values[:#{name}] || FROZEN_EMPTY_ARRAY #   @values[:select] || []
+          else
+            @values[:#{name}] || []
+          end
         end                                       # end
                                                   #
         def #{name}_values=(values)               # def select_values=(values)
@@ -100,7 +104,11 @@ module ActiveRecord
 
     FROZEN_EMPTY_HASH = {}.freeze
     def create_with_value # :nodoc:
+      if $set_hash
         @values[:create_with] || FROZEN_EMPTY_HASH
+      else
+        @values[:create_with] || {}
+      end
     end
 
     alias extensions extending_values
diff --git a/activerecord/lib/active_record/relation/where_clause.rb b/activerecord/lib/active_record/relation/where_clause.rb
index f6613e4..8e8e2c1 100644
--- a/activerecord/lib/active_record/relation/where_clause.rb
+++ b/activerecord/lib/active_record/relation/where_clause.rb
@@ -81,12 +81,16 @@ module ActiveRecord
       end
 
       def self.empty
+        if $set_clause
           @empty ||= begin
             empty = new([], [])
             # This method memoizes, which will blow up on a frozen object
             empty.send(:referenced_columns)
             empty.freeze
           end
+        else
+          new([], [])
+        end
       end
 
       protected
```
Results:
```
Calculating -------------------------------------
spawn: set_array: false, set_clause: false, set_hash: false
                        28.743k i/100ms
spawn: set_array: true, set_clause: false, set_hash: false
                        25.634k i/100ms
spawn: set_array: false, set_clause: true, set_hash: false
                        22.658k i/100ms
spawn: set_array: false, set_clause: false, set_hash: true
                        26.748k i/100ms
spawn: set_array: true, set_clause: true, set_hash: false
                        26.060k i/100ms
spawn: set_array: true, set_clause: false, set_hash: true
                        26.109k i/100ms
spawn: set_array: false, set_clause: true, set_hash: true
                        27.297k i/100ms
spawn: set_array: true, set_clause: true, set_hash: true
                        24.451k i/100ms
-------------------------------------------------
spawn: set_array: false, set_clause: false, set_hash: false
                        525.659k (± 6.3%) i/s -      2.644M
spawn: set_array: true, set_clause: false, set_hash: false
                        450.370k (± 8.2%) i/s -      2.256M
spawn: set_array: false, set_clause: true, set_hash: false
                        462.342k (± 7.8%) i/s -      2.311M
spawn: set_array: false, set_clause: false, set_hash: true
                        440.271k (± 9.9%) i/s -      2.193M
spawn: set_array: true, set_clause: true, set_hash: false
                        443.915k (±12.1%) i/s -      2.189M
spawn: set_array: true, set_clause: false, set_hash: true
                        447.401k (± 7.5%) i/s -      2.245M
spawn: set_array: false, set_clause: true, set_hash: true
                        470.791k (± 6.6%) i/s -      2.348M
spawn: set_array: true, set_clause: true, set_hash: true
                        477.696k (± 4.4%) i/s -      2.396M

Comparison:
spawn: set_array: false, set_clause: false, set_hash: false:   525659.4 i/s
spawn: set_array: true, set_clause: true, set_hash: true:   477696.1 i/s - 1.10x slower
spawn: set_array: false, set_clause: true, set_hash: true:   470791.4 i/s - 1.12x slower
spawn: set_array: false, set_clause: true, set_hash: false:   462342.2 i/s - 1.14x slower
spawn: set_array: true, set_clause: false, set_hash: false:   450370.3 i/s - 1.17x slower
spawn: set_array: true, set_clause: false, set_hash: true:   447401.4 i/s - 1.17x slower
spawn: set_array: true, set_clause: true, set_hash: false:   443915.3 i/s - 1.18x slower
spawn: set_array: false, set_clause: false, set_hash: true:   440271.2 i/s - 1.19x slower

Calculating -------------------------------------
merge: set_array: false, set_clause: false, set_hash: false
                         1.958k i/100ms
merge: set_array: true, set_clause: false, set_hash: false
                         1.965k i/100ms
merge: set_array: false, set_clause: true, set_hash: false
                         1.979k i/100ms
merge: set_array: false, set_clause: false, set_hash: true
                         1.995k i/100ms
merge: set_array: true, set_clause: true, set_hash: false
                         1.997k i/100ms
merge: set_array: true, set_clause: false, set_hash: true
                         1.986k i/100ms
merge: set_array: false, set_clause: true, set_hash: true
                         1.976k i/100ms
merge: set_array: true, set_clause: true, set_hash: true
                         1.971k i/100ms
-------------------------------------------------
merge: set_array: false, set_clause: false, set_hash: false
                         20.834k (± 4.3%) i/s -    105.732k
merge: set_array: true, set_clause: false, set_hash: false
                         20.933k (± 3.2%) i/s -    106.110k
merge: set_array: false, set_clause: true, set_hash: false
                         19.273k (± 7.0%) i/s -     96.971k
merge: set_array: false, set_clause: false, set_hash: true
                         19.175k (± 7.3%) i/s -     95.760k
merge: set_array: true, set_clause: true, set_hash: false
                         19.283k (± 7.3%) i/s -     97.853k
merge: set_array: true, set_clause: false, set_hash: true
                         19.493k (± 5.3%) i/s -     97.314k
merge: set_array: false, set_clause: true, set_hash: true
                         19.517k (± 5.7%) i/s -     98.800k
merge: set_array: true, set_clause: true, set_hash: true
                         19.850k (± 5.9%) i/s -    100.521k

Comparison:
merge: set_array: true, set_clause: false, set_hash: false:    20933.3 i/s
merge: set_array: false, set_clause: false, set_hash: false:    20833.9 i/s - 1.00x slower
merge: set_array: true, set_clause: true, set_hash: true:    19849.9 i/s - 1.05x slower
merge: set_array: false, set_clause: true, set_hash: true:    19516.6 i/s - 1.07x slower
merge: set_array: true, set_clause: false, set_hash: true:    19493.4 i/s - 1.07x slower
merge: set_array: true, set_clause: true, set_hash: false:    19283.4 i/s - 1.09x slower
merge: set_array: false, set_clause: true, set_hash: false:    19272.8 i/s - 1.09x slower
merge: set_array: false, set_clause: false, set_hash: true:    19175.3 i/s - 1.09x slower
```

**Comment by tamird** [comment]
Well this is bizarre. If I run the benchmarks in reverse (with the caching on first), the results are reversed (caching is fastest). Any ideas?

**Comment by tamird** [comment]
Closing in favour of #22550.
